File: v32_visitor/java/ShoppingCart.java
package v32_visitor.java;

import java.util.ArrayList;
import java.util.List;

interface Item {
    void accept(Visitor visitor);
}

class Book implements Item {
    private double price;

    public Book(double price) {
        this.price = price;
    }

    public double getPrice() {
        return price;
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

class Electronics implements Item {
    private double price;

    public Electronics(double price) {
        this.price = price;
    }

    public double getPrice() {
        return price;
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

interface Visitor {
    void visit(Book book);
    void visit(Electronics electronics);
}

class RegularPriceVisitor implements Visitor {
    private double totalCost = 0;

    @Override
    public void visit(Book book) {
        totalCost += book.getPrice();
    }

    @Override
    public void visit(Electronics electronics) {
        totalCost += electronics.getPrice();
    }

    public double getTotalCost() {
        return totalCost;
    }
}

class DiscountPriceVisitor implements Visitor {
    private double totalCost = 0;

    @Override
    public void visit(Book book) {
        totalCost += book.getPrice() * 0.8; // 20% discount on books
    }

    @Override
    public void visit(Electronics electronics) {
        totalCost += electronics.getPrice() * 0.9; // 10% discount on electronics
    }

    public double getTotalCost() {
        return totalCost;
    }
}

public class ShoppingCart {
    public static void main(String[] args) {
        List<Item> items = new ArrayList<>();
        items.add(new Book(50));
        items.add(new Book(30));
        items.add(new Electronics(100));

        RegularPriceVisitor regularVisitor = new RegularPriceVisitor();
        DiscountPriceVisitor discountVisitor = new DiscountPriceVisitor();

        for (Item item : items) {
            item.accept(regularVisitor);
            item.accept(discountVisitor);
        }

        System.out.println("Regular total cost: " + regularVisitor.getTotalCost());
        System.out.println("Discount total cost: " + discountVisitor.getTotalCost());
    }
}

File: v32_visitor/python/shopping_cart.py
from abc import ABC, abstractmethod

class Item(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass

class Book(Item):
    def __init__(self, price):
        self.price = price

    def get_price(self):
        return self.price

    def accept(self, visitor):
        visitor.visit_book(self)

class Electronics(Item):
    def __init__(self, price):
        self.price = price

    def get_price(self):
        return self.price

    def accept(self, visitor):
        visitor.visit_electronics(self)

class Visitor(ABC):
    @abstractmethod
    def visit_book(self, book):
        pass

    @abstractmethod
    def visit_electronics(self, electronics):
        pass

class RegularPriceVisitor(Visitor):
    def __init__(self):
        self.total_cost = 0

    def visit_book(self, book):
        self.total_cost += book.get_price()

    def visit_electronics(self, electronics):
        self.total_cost += electronics.get_price()

    def get_total_cost(self):
        return self.total_cost

class DiscountPriceVisitor(Visitor):
    def __init__(self):
        self.total_cost = 0

    def visit_book(self, book):
        self.total_cost += book.get_price() * 0.8  # 20% discount on books

    def visit_electronics(self, electronics):
        self.total_cost += electronics.get_price() * 0.9  # 10% discount on electronics

    def get_total_cost(self):
        return self.total_cost

def main():
    items = [Book(50), Book(30), Electronics(100)]

    regular_visitor = RegularPriceVisitor()
    discount_visitor = DiscountPriceVisitor()

    for item in items:
        item.accept(regular_visitor)
        item.accept(discount_visitor)

    print("Regular total cost: ", regular_visitor.get_total_cost())
    print("Discount total cost: ", discount_visitor.get_total_cost())

if __name__ == "__main__":
    main()

